File: 2024/problems/tarotpoker/submissions/partially_accepted/no_joker_handling_80.cpp
#include "bits/stdc++.h"

using namespace std;

vector<long long int> ranks[20];

int cnt = 0;
string cards[78];
vector<string> suits = {"S", "B", "V", "M"};

int get_value(int h) {
  if (h < 56) return h / 4;
  else return h;
}

int get_comp_value(int h) {
  if (h < 56) {
    if (h / 4 < 10) return h / 4 + 1;
    else return h / 4 + 20; // to not mix with high cards
  } else return h - 56;
}

int get_suit(int h) {
  if (h < 56) return h % 4;
  else return h;
}

tuple<vector<int>, vector<int>, vector<int>> get_values_and_suits(const vector<int>& h) {
  const auto N = h.size();
  // values
  vector<int> v(N);
  for (size_t i = 0; i < N; ++i) v[i] = get_value(h[i]);
  sort(v.begin(), v.end());

  // comparable values
  vector<pair<int, int>> c_(N);
  for (size_t i = 0; i < N; ++i) c_[i] = {get_value(h[i]), get_comp_value(h[i])};
  sort(c_.begin(), c_.end()); // sort according to value
  vector<int> c(N);
  for (size_t i = 0; i < N; ++i) c[i] = c_[i].second;

  // suits
  vector<int> s(N);
  for (size_t i = 0; i < N; ++i) s[i] = get_suit(h[i]);
  sort(s.begin(), s.end());

  return {v, c, s};
}

int solve2(const vector<int>& h) {
  auto [v, c, s] = get_values_and_suits(h);

  if (v[0] < 10 && v[1] >= 14 && c[0] == c[1]) return 15; // comp pair of low and high
  if (v[0] == v[1]) return 18; // 2 same value
  return 100;
}

int solve3(const vector<int>& h) {
  auto [v, c, s] = get_values_and_suits(h);

  if (v[0] == v[2]) return 16; // 3 same value
  return 100;
}

int solve4(const vector<int>& h) {
  auto [v, c, s] = get_values_and_suits(h);

  if (v[2] < 14 && v[3] >= 14 && c[0] + c[1] + c[2] == c[3]) return 7; // 3 low, 1 high comp to sum of low
  if (v[0] == v[3]) return 9; // 4 same value
  if (v[1] < 10 && v[2] >= 14 && c[0] == c[2] && c[1] == c[3]) return 13; // 2 low (numbers), 2 high, each high comp to corresponding low
  if (v[0] == v[1] && v[2] < 10 && v[3] >= 14 && c[2] == c[3]) return 14; // 2 same value, 1 low (number) comp to remaining high
  if (v[1] == v[2] && v[0] < 10 && v[3] >= 14 && c[0] == c[3]) return 14; //   does not matter which are same value, the combination set is the same
  if (v[2] == v[3] && v[0] < 10 && v[1] >= 14 && c[0] == c[1]) return 14;
  if (v[0] == v[1] && v[2] == v[3]) return 17; // 2 pairs of same values
  return 100;
}

// result and best combination, sorted in descending order
pair<int, vector<int>> solve5(const vector<int>& h) {
  auto [v, c, s] = get_values_and_suits(h);

  // solve subsets of 2 (for case 18)
  pair<int, vector<int>> result2 = {100, {}};
  for (int i = 0; i < 4; ++i) {
    for (int j = i + 1; j < 5; ++j) {
      vector<int> subh = {h[i], h[j]};
      auto result = solve2(subh);
      if (result < result2.first) { // subh are in decreasing order, so only strictly better result
        result2 = {result, subh};
      }
    }
  }
  // solve subsets of 3 (for case 16)
  pair<int, vector<int>> result3 = {100, {}};
  for (int i = 3; i >= 0; i--) {
    for (int j = 4; j > i; j--) {
      vector<int> subh = h; subh.erase(subh.begin() + j); subh.erase(subh.begin() + i); // erase order!
      auto result = solve3(subh);
      if (result != 100) { // subh are in decreasing order, only 1 possible result
        result3 = {result, subh}; break;
      }
    }
    if (result3.first != 100) break;
  }
  // solve subsets of 4 (for cases 7, 9, 13, 14, 17)
  pair<int, vector<int>> result4 = {100, {}};
  for (int i = 4; i >= 0; i--) {
    vector<int> subh = h; subh.erase(subh.begin() + i);
    auto result = solve4(subh);
    if (result < result4.first) { // subh are in decreasing order, so only strictly better result
      result4 = {result, subh};
    }
  }

  if (v[0] == v[4]) return {1, h}; // 5 same value
  if (v[0] >= 14) return {2, h}; // 5 high cards
  if (v[4] < 14 && s[0] == s[4] && v[0] + 1 == v[1] && v[1] + 1 == v[2] && v[2] + 1 == v[3] && v[3] + 1 == v[4]) return {3, h}; // 5 (low) same suit, sequential
  if (v[3] < 14 && v[4] >= 14 && s[0] != s[1] && s[1] != s[2] && s[2] != s[3] && s[3] != s[4]) return {4, h}; // 4 low, 1 high, all(!) different suits
  if (v[3] < 14 && v[4] >= 14 && c[0] + c[1] + c[2] + c[3] == c[4]) return {5, h}; // 4 low, 1 high, sum of low compares to high
  if (v[3] < 14 && v[4] >= 14 && c[0] == c[4] && c[0] == c[3]) return {6, h}; // 4 low, 1 high, each low compares to high
  if (result4.first == 7) return result4;
  if (v[0] >= 10 && v[4] < 14) return {8, h}; // 5 low, no numbers
  if (result4.first == 9) return result4;
  if ((v[0] == v[2] && v[3] == v[4]) || (v[0] == v[1] && v[2] == v[4])) return {10, h}; // 3+2 or 2+3 same value
  if (v[4] < 14 && s[0] == s[4]) return {11, h}; // 5 (low), same suit
  if (v[0] + 1 == v[1] && v[1] + 1 == v[2] && v[2] + 1 == v[3] && v[3] + 1 == v[4]) return {12, h}; // 5 sequential
  if (result4.first == 13) return result4;
  if (result4.first == 14) return result4;
  if (result2.first == 15) return result2;
  if (result3.first == 16) return result3;
  if (result4.first == 17) return result4;
  if (result2.first == 18) return result2; // 2 same value
  if (v[4] < 14) return {19, h}; // 5 low cards
  return {20, {h[0]}}; // highest card
}

pair<int, vector<int>> solve(vector<int> h) {
  auto result = solve5(h);
  // remove winning combination from the hand
  for (int i = 0; i < result.second.size(); ++i) {
    for (int j = 0; j < h.size(); ++j) {
      if (h[j] == result.second[i]) {
        h.erase(h.begin() + j); break;
      }
    }
  }
  // augment winning combination with the remaining cards for tiebreak
  for (auto x : h) result.second.push_back(x);
  return result;
}

vector<int> get_hand(vector<string> s) {
  vector<int> h(5);
  for (int i = 0; i < 5; ++i)
    for (int j = 0; j < cnt; ++j) if (cards[j] == s[i]) h[i] = j;
  sort(h.begin(), h.end(), std::greater<int>());
  return h;
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);

  for (int i = 1; i <= 10; ++i)
    for (auto suit : suits) cards[cnt++] = suit + to_string(i);
  for (auto court : {"G", "R", "D", "K"})
    for (auto suit : suits) cards[cnt++] = suit + court;
  cards[cnt++] = "O";
  cards[cnt++] = "I";
  cards[cnt++] = "II";
  cards[cnt++] = "III";
  cards[cnt++] = "IV";
  cards[cnt++] = "V";
  cards[cnt++] = "VI";
  cards[cnt++] = "VII";
  cards[cnt++] = "VIII";
  cards[cnt++] = "IX";
  cards[cnt++] = "X";
  cards[cnt++] = "XI";
  cards[cnt++] = "XII";
  cards[cnt++] = "XIII";
  cards[cnt++] = "XIV";
  cards[cnt++] = "XV";
  cards[cnt++] = "XVI";
  cards[cnt++] = "XVII";
  cards[cnt++] = "XVIII";
  cards[cnt++] = "XIX";
  cards[cnt++] = "XX";
  cards[cnt++] = "XXI";

  vector<string> konrads(5);
  for (int i = 0; i < 5; ++i) cin >> konrads[i];
  auto konradh = get_hand(konrads);
  vector<string> atlis(5);
  for (int i = 0; i < 5; ++i) cin >> atlis[i];
  auto atlih = get_hand(atlis);

  auto konrad = solve(konradh);
  auto atli = solve(atlih);
  cout << konrad.first << ' ' << atli.first << endl;
  // order scores same as winning combinations
  konrad.first = -konrad.first;
  atli.first = -atli.first;
  if (konrad > atli)
    cout << "Konrad" << endl;
  if (konrad < atli)
    cout << "Atli" << endl;
  if (konrad == atli)
    cout << "Jafntefli" << endl;
  return 0;
}

File: 2024/problems/tarotpoker/submissions/partially_accepted/no_joker_handling_80.py
#!/usr/bin/python3

card_to_ord, suit, value, parallel_value = dict(), dict(), dict(), dict()

high_cards = ["O", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "XVII", "XVIII", "XIX", "XX", "XXI"]
highs = dict()
for i in range(len(high_cards)):
    highs[high_cards[i]] = i
    card_to_ord[high_cards[i]] = 57 + i
    suit[high_cards[i]] = None
    value[high_cards[i]] = i + 15
    parallel_value[high_cards[i]] = i

all_cards = high_cards[:]
lows = dict()
for (i, sort) in enumerate(["S", "B", "V", "M"]):
    for (j, val) in enumerate(["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "G", "R", "D", "K"]):
        lows[sort + val] = 4 * j + i
        card_to_ord[sort + val] = 4 * j + i + 1
        suit[sort + val] = sort
        value[sort + val] = j + 1
        if val in "GRDK":
            parallel_value[sort + val] = None
        else:
            parallel_value[sort + val] = j + 1
        all_cards.append(sort + val)

def make_row(hand):
    if len(set(hand)) != len(hand):
        return False
    ord_vals = [value[card] for card in hand]
    if len(set(ord_vals)) != len(ord_vals):
        return False
    if min(ord_vals) < 15 and max(ord_vals) >= 15:
        return False
    return min(ord_vals) + len(hand) - 1 == max(ord_vals)

def prep_return(rank, pat, lft):
    pat_ord = sorted([card_to_ord[card] for card in pat], reverse=True)
    lft_ord = sorted([card_to_ord[card] for card in lft], reverse=True)
    return [rank] + pat_ord + lft_ord

def rank_hand(hand):
    values = [value[card] for card in hand]
    lows_para = [parallel_value[card] for card in hand if card in lows]
    high_para = [parallel_value[card] for card in hand if card in highs]
    suits = [suit[card] for card in hand]
    lows_vals = [value[card] for card in hand if card in lows]
    high_vals = [value[card] for card in hand if card in highs]
    vcounts = [values.count(values[i]) for i in range(5)]
    if len(set(values)) == 1:
        return prep_return(20, hand, [])
    if len(lows_vals) == 0:
        return prep_return(19, hand, [])
    if len(high_vals) == 0 and make_row(hand) and len(set(suits)) == 1:
        return prep_return(18, hand, [])
    if len(set(suits)) == 5:
        return prep_return(17, hand, [])
    if len(high_vals) == 1 and len(lows_vals) == 4 and None not in lows_para and sum(lows_para) == sum(high_para):
        return prep_return(16, hand, [])
    if len(high_vals) == 1 and len(set(lows_vals)) == 1 and lows_para[0] == high_para[0]:
        return prep_return(15, hand, [])
    if len(lows_para) >= 3 and len(high_para) >= 1:
        mn, ind = 1000, -1
        for i in range(5):
            new_lows_para = [parallel_value[hand[j]] for j in range(5) if j != i and hand[j] in lows]
            new_high_para = [parallel_value[hand[j]] for j in range(5) if j != i and hand[j] in highs]
            if None in new_lows_para or len(new_lows_para) != 3 or len(new_high_para) != 1:
                continue
            if sum(new_lows_para) != sum(new_high_para):
                continue
            x = card_to_ord[hand[i]]
            if x < mn:
                mn = x
                ind = i
        if mn != 1000:
            return prep_return(14, [hand[i] for i in range(5) if i != ind], [hand[ind]])
    if all([card[-1] in "GRDK" for card in hand]):
        return prep_return(13, hand, [])
    if len(set(values)) == 2 and sorted(vcounts) == [1, 4, 4, 4, 4]:
        ind = vcounts.index(1)
        return prep_return(12, [hand[j] for j in range(5) if j != ind], [hand[ind]])
    if len(set(values)) == 2: # Can't be 4, 1
        return prep_return(11, hand, [])
    if None not in suits and len(set(suits)) == 1:
        return prep_return(10, hand, [])
    if make_row(hand):
        return prep_return(9, hand, [])
    if len(high_vals) >= 2 and len(lows_vals) >= 2:
        for i in range(5):
            if len(high_vals) - (1 if hand[i] in highs else 0) != 2:
                continue
            new_lows_para = [parallel_value[hand[j]] for j in range(5) if j != i and hand[j] in lows]
            new_high_para = sorted([parallel_value[hand[j]] for j in range(5) if j != i and hand[j] in highs])
            if None in new_lows_para:
                continue
            new_lows_para.sort()
            if new_lows_para == new_high_para:
                return prep_return(8, [hand[j] for j in range(5) if j != i], [hand[i]])
    if len(high_vals) >= 1 and len(lows_vals) >= 1:
        mn, ind = 1000, -1
        for i in range(5):
            new_lows_para = [parallel_value[hand[j]] for j in range(5) if j != i and hand[j] in lows]
            new_high_para = sorted([parallel_value[hand[j]] for j in range(5) if j != i and hand[j] in highs])
            if len(new_lows_para) not in [1, 3]:
                continue
            if len(new_high_para) not in [1, 3]:
                continue
            if len(new_high_para) == len(new_lows_para):
                continue
            if len(new_high_para) == 3:
                if new_lows_para[0] not in new_high_para:
                    continue
                del new_high_para[new_high_para.index(new_lows_para[0])]
                if len(set(new_high_para)) != 1:
                    continue
            else:
                if new_high_para[0] not in new_lows_para:
                    continue
                del new_lows_para[new_lows_para.index(new_high_para[0])]
                if len(set(new_lows_para)) != 1:
                    continue
                if new_lows_para[0] is None:
                    actual_val = [value[hand[j]] for j in range(5) if j != i and hand[j] in lows and value[hand[j]] >= 11]
                    if len(set(actual_val)) > 1:
                        continue
            x = card_to_ord[hand[i]]
            if x < mn:
                mn = x
                ind = i
        if mn != 1000:
            return prep_return(7, [hand[j] for j in range(5) if j != ind], [hand[ind]])
    if len(high_vals) >= 1 and len(lows_vals) >= 1: # must make sure to pick maximum corresp.
        mx, ind = -1, -1
        for i in range(5): # unique i because no 2 high pair
            if hand[i] not in highs:
                continue
            if parallel_value[hand[i]] not in lows_para:
                continue
            for j in range(5):
                if hand[j] not in lows:
                    continue
                if parallel_value[hand[i]] != parallel_value[hand[j]]:
                    continue
                x = card_to_ord[hand[j]]
                if x > mx:
                    mx = x
                    ind = (i, j)
        if mx != -1:
            return prep_return(6, [hand[j] for j in ind], [hand[j] for j in range(5) if j not in ind])
    for i in range(5):
        if values.count(values[i]) == 3:
            return prep_return(5, [hand[j] for j in range(5) if values[j] == values[i]], [hand[j] for j in range(5) if values[i] != values[j]])
    if len(set(values)) == 3: # Not 3, 1, 1 so must be 2, 2, 1
        return prep_return(4, [hand[j] for j in range(5) if values.count(values[j]) == 2], [hand[j] for j in range(5) if values.count(values[j]) == 1])
    if len(set(values)) < 5:
        inds = [i for i in range(5) if values.count(values[i]) == 2]
        return prep_return(3, [hand[i] for i in inds], [hand[i] for i in range(5) if i not in inds])
    if len(high_vals) == 0:
        return prep_return(2, hand, [])
    return prep_return(1, hand, [])

if __name__ == "__main__":
    konrad = rank_hand(list(input().strip().replace("J", "XXI").split()))
    atli = rank_hand(list(input().strip().replace("J", "XXI").split()))
    print(21 - konrad[0], 21 - atli[0])
    if konrad > atli:
        print("Konrad")
    elif atli > konrad:
        print("Atli")
    else:
        print("Jafntefli")
